cmd/webhook/start.go

```diff
@@ -0,0 +1,210 @@
+		"Secure port that the webhook listens on")
+}
+
+// serve handles the http portion of a request prior to handing to an admit
+// function
+func serve(w http.ResponseWriter, r *http.Request, admit admitHandler) {
+	var body []byte
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
+	}
+
+	// verify the content type is accurate
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/json" {
+		glog.Errorf("contentType=%s, expect application/json", contentType)
+		return
+	}
+
+	glog.V(2).Info(fmt.Sprintf("handling request: %s", body))
+
+	deserializer := webhook.Codecs.UniversalDeserializer()
+	obj, gvk, err := deserializer.Decode(body, nil, nil)
+	if err != nil {
+		msg := fmt.Sprintf("Request could not be decoded: %v", err)
+		glog.Error(msg)
+		http.Error(w, msg, http.StatusBadRequest)
+		return
+	}
+
+	var responseObj runtime.Object
+	switch *gvk {
+	case v1beta1.SchemeGroupVersion.WithKind("AdmissionReview"):
+		requestedAdmissionReview, ok := obj.(*v1beta1.AdmissionReview)
+		if !ok {
+			glog.Errorf("Expected v1beta1.AdmissionReview but got: %T", obj)
+			return
+		}
+		responseAdmissionReview := &v1beta1.AdmissionReview{}
+		responseAdmissionReview.SetGroupVersionKind(*gvk)
+		responseAdmissionReview.Response = admit.v1beta1(*requestedAdmissionReview)
+		responseAdmissionReview.Response.UID = requestedAdmissionReview.Request.UID
+		responseObj = responseAdmissionReview
+	default:
+		msg := fmt.Sprintf("Unsupported group version kind: %v", gvk)
+		glog.Error(msg)
+		http.Error(w, msg, http.StatusBadRequest)
+		return
+	}
+
+	respBytes, err := json.Marshal(responseObj)
+	if err != nil {
+		glog.Error(err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	glog.V(2).Info(fmt.Sprintf("sending response: %s", string(respBytes[:])))
+	w.Header().Set("Content-Type", "application/json")
+	if _, err := w.Write(respBytes); err != nil {
+		glog.Error(err)
+	}
+}
+
+func serveMutateCustomResource(w http.ResponseWriter, r *http.Request) {
+	serve(w, r, newDelegateToV1beta1AdmitHandler(webhook.MutateCustomResource))
+}
+
+func serveValidateCustomResource(w http.ResponseWriter, r *http.Request) {
+	serve(w, r, newDelegateToV1beta1AdmitHandler(webhook.ValidateCustomResource))
+}
+
+func newDelegateToV1beta1AdmitHandler(f admitv1beta1Func) admitHandler {
+	return admitHandler{
+		v1beta1: f,
+	}
+}
+
+func runStartCmd(cmd *cobra.Command, args []string) {
+	if err := webhook.SetupInClusterClient(); err != nil{
+		glog.Error(err)
+		panic(err)
+	}
+
+	keyPair, err := webhook.NewTlsKeypairReloader(certFile, keyFile)
+	if err != nil {
+		glog.Fatalf("error load certificate: %s", err.Error())
+	}
+
+
+	http.HandleFunc("/mutating-custom-resource", serveMutateCustomResource)
+	http.HandleFunc("/validating-custom-resource", serveValidateCustomResource)
+	http.HandleFunc("/readyz", func(w http.ResponseWriter, req *http.Request) { w.Write([]byte("ok")) })
```

Comment:
what's this handler used for?

Reply:
This is for `readinessProbe` of the pod, which can be used to detect if a pod is ready for service.